torch/nn/modules/module.py

```diff
@@ -56,6 +70,7 @@ def __init__(self):
         self._forward_pre_hooks = OrderedDict()
         self._modules = OrderedDict()
         self.training = True
+        self._name = ''
```

Comment:
Because we have added a new field to the module, I wonder if we shouldn't adapt the `__getstate__` to recreate the missing fields for backward-compatibility reasons. @ezyang what do you think?

Reply:
I don't have very much experience with Python's pickler, so I can't say for certain, but it is definitely better for people's old pickles to keep unpickling after this change. I don't know what is necessary to actually make this happen.

Reply:
Unpickling will still work, but the `name_` attribute won't be present, so the code will fail. I'm adding a
```
    def __setstate__(self, d):
        self.__dict__.update(d)
        self.__dict__.setdefault('name_', '')
```
to Module so that the field will still exist after unpickling an old pickle. I don't think we need a `__getstate__`, but I may be missing something.

Reply:
Also: we could set the default name to something unique. Do you have ideas?

Reply:
Yeah, I think we just need to modify the deserialization part (`__setstate__`, I think I confused the names :-) ).
I think the only situation where the `_name` won't be assigned to a name is in the main model class, all others will have names, right? In this case, I think we could name it something like `model`?

Reply:
Thanks @fmassa , I have modified `__setstate__` in the last commit.
I was especially referring to unpickling of old pickles in `__setstate__`. Right now it will just make sure that `_name` exists, but it will set it to the empty string. This will lead to incorrect grouping of nodes for older models, if one were to, say, display a pre-trained network.

Using `id(self)` in these cases might be ok, although it's not stable in subsequent unpicklings.
Alternatively we have to assign a progressive number to each unpicked object of a given class, although this is a bit more involved.

What do you think?

Reply:
Ah, I misunderstood what you meant.
It is possible to traverse the structure the modules and assign `_name` to each of the modules, so that the new behaviour can be found in old serialized models as well.
Here is an approximate snippet:
```python
for name, module in self.named_modules():
    module._name = name
```
What do you think?

Reply:
(note: `named_modules` returns each module only once, so the namings might not be consistent because the name of a module would be the last assigned one during model creation, but the first one during deserialization)

Reply:
(we should also note that such kind of problems would only happen if the user serialized the whole `Module` object, and not if he serialized the `state_dict`, which is the preferred way to do it)

Reply:
What about we use `named_children` in `__setstate__`? Since all instances of modules will be unpickled and children will have to be unpickled first for the unpickling to succeed, at the end all modules will be named (except the top-most, which we don't really care about)

Reply:
`__setstate__` would become:
```
    def __setstate__(self, state):
        self.__dict__.update(state)
        if '_name' not in self.__dict__:
            self._name = ''
            for name, module in self.named_children():
                module._name = name
```

Reply:
Looks good to me as well.